src/Microsoft.Restier.Core/Api.cs

```diff
@@ -78,16 +78,40 @@ public static class Api // TODO GitHubIssue#25,#26 : transactions, exception fil
         {
             Ensure.NotNull(context, "context");
             var config = context.Configuration;
-            if (config.Model == null)
+            if (config.Model != null)
             {
-                var builder = context.GetApiService<IModelBuilder>();
-                if (builder != null)
-                {
-                    config.Model = await builder.GetModelAsync(new InvocationContext(context), cancellationToken);
-                }
+                return config.Model;
+            }
+
+            var builder = context.GetApiService<IModelBuilder>();
+            if (builder == null)
+            {
+                return null;
+            }
+
+            Task<IEdmModel> running;
+            var source = config.CompeteModelGeneration(out running);
+            if (source == null)
+            {
+                return await running;
             }
 
-            return config.Model;
+            try
+            {
+                var model = await builder.GetModelAsync(new InvocationContext(context), cancellationToken);
+                source.SetResult(model);
+                return model;
+            }
+            catch (AggregateException e)
```

Comment:
Are we going to do anything special for an `AggregateException` in the future? If not, we may consider merge the two catch-clauses.